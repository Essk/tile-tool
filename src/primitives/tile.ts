import uniqid from 'uniqid';
import { FileEntity } from './Storeable';

type Opts = {
    name?: string;
    id?: string;
    px?: number[];
    paletteSet?: string;
    currentPalette?: any;
    defaultPalette?: any;
};
export type TileFileData = {
    _name: string;
    _id: string;
    _px: number[];
    _paletteSet: string;
    _currentPalette: any;
    _defaultPalette: any;
};
export class Tile extends FileEntity {
    public static hydrate(tileFile: TileFileData) {
        return new Tile(tileFile);
    }
    private _name: string;
    private _id: string;
    private _px: number[];
    private _len: number;
    private _paletteSet: string;
    private _currentPalette: number;
    private _defaultPalette: string;
    constructor(opts: Opts | TileFileData) {
        super();
        if (this.isFromFile(opts)) {
            this._name = opts._name;
            this._id = opts._id;
            this._len = 64;
            this._px = opts._px ;
            this._paletteSet = opts._paletteSet;
            this._currentPalette = opts._currentPalette;
            this._defaultPalette = opts._defaultPalette;
        } else {
            this._name = opts.name || 'new tile';
            this._id = opts.id || this.makeId();
            this._len = 64;
            this._px = opts.px ? this.makePx(this._len, opts.px) : this.makePx(this._len);
            this._paletteSet = opts.paletteSet || '';
            this._currentPalette =  this.validPalette(opts.currentPalette) ? opts.currentPalette : 0;
            this._defaultPalette = opts.defaultPalette || '';
        }
    }
    public updatePx(idx: number, value: number) {
        if (this.validIdx(idx) && this.validValue(value)) {
            this.px[idx] = value;
        }
    }
    private makeId(): string {
        return uniqid('tile-');
    }
    private makePx(len: number, px?: number[]): number[] {
        const _px = px || [];
        if (_px.length >= 0 && _px.length <= len) {
            for (let i = _px.length; i < len; i++) {
                _px.push(0);
            }
        } else if (_px.length > len) {
            return _px.splice(len);
        }
        return _px;
    }
    get px(): number[] {
        return this._px;
    }
    set px(newPx: number[]) {
        this._px = newPx;
    }
    get name(): string {
        return this._name;
    }
    get id(): string {
        return this._id;
    }
    get paletteSet(): string {
        return this._paletteSet;
    }
    set palette_set(paletteSet: string) {
        this._paletteSet = paletteSet;
    }
    get currentPalette(): number {
        return this._currentPalette;
    }

    set currentPalette(idx: number) {
        this._currentPalette = this.validPalette(idx) ? idx : 0 ;
    }
    get defaultPalette(): string {
        return this._defaultPalette;
    }
    set defaultPalette(paletteId: string) {
        this._defaultPalette = paletteId;
    }
    private validPalette(idx: number ): boolean {
        if (idx === undefined) {
            return false;
        }
        return Number.isInteger(idx)
        && idx >= 0
        && idx < 16;
    }
    private validIdx(idx: number): boolean {
        return Number.isInteger(idx) && idx >= 0 && idx < this._len;
    }
    private validValue(value: number): boolean {
        return Number.isInteger(value) && value >= 0 && value < 16;
    }
    private isFromFile(opts: Opts | TileFileData): opts is TileFileData {
        return (opts as TileFileData)._id !== undefined;
    }

}
