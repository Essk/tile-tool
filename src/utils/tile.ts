import uniqid from 'uniqid';

type Opts = {
    name?: string;
    id?: string;
    px?: number[];
    paletteSet?: string;
    currentPalette?: any;
};
type TileFileData = {
    _name: string;
    _id: string;
    _px: number[];
    _paletteSet: string;
    _currentPalette: any;
};
export class Tile {
    private _name: string;
    private _id: string;
    private _px: number[];
    private _len: number;
    private _paletteSet: string;
    private _currentPalette: number;
    constructor(opts: Opts | TileFileData) {
        if (this.isFromFile(opts)) {
            this._name = opts._name;
            this._id = opts._id;
            this._len = 64;
            this._px = opts._px ;
            this._paletteSet = opts._paletteSet;
            this._currentPalette = opts._currentPalette;
        } else {
            this._name = opts.name || 'new tile';
            this._id = opts.id || this.makeId();
            this._len = 64;
            this._px = opts.px ? this.makePx(this._len, opts.px) : this.makePx(this._len) ;
            this._paletteSet = opts.paletteSet || '';
            this._currentPalette =  this.validPalette(opts.currentPalette) ? opts.currentPalette : 0;
        }
    }
    public updatePx(idx: number, value: number) {
        if (this.validIdx(idx) && this.validValue(value)) {
            this.px[idx] = value;
        }
    }
    private makeId(): string {
        return uniqid('tile-');
    }
    private makePx(len: number, px?: number[]): number[] {
        const _px = px || [];
        if (_px.length > 0 && _px.length <= len) {
            for (let i = _px.length; i < len; i++) {
                _px.push(0);
            }
        } else if (_px.length > len) {
            return _px.splice(len);
        }
        return _px;
    }
    get px(): number[] {
        return this._px;
    }
    set px(newPx: number[]) {
        this._px = newPx;
    }
    get name(): string {
        return this._name;
    }
    get id(): string {
        return this._id;
    }
    get paletteSet(): string {
        return this._paletteSet;
    }
    set palette_set(paletteSet: string) {
        this._paletteSet = paletteSet;
    }
    get currentPalette(): number {
        return this._currentPalette;
    }

    set currentPalette(idx: number) {
        this._currentPalette = this.validPalette(idx) ? idx : 0 ;
    }

    private validPalette(idx: number ): boolean {
        if (idx === undefined) {
            return false;
        }
        return Number.isInteger(idx)
        && idx >= 0
        && idx < 16;
    }
    private validIdx(idx: number): boolean {
        return Number.isInteger(idx) && idx >= 0 && idx < this._len;
    }
    private validValue(value: number): boolean {
        return Number.isInteger(value) && value >= 0 && value < 16;
    }
    private isFromFile(opts: Opts | TileFileData): opts is TileFileData {
        return (opts as TileFileData)._id !== undefined;
    }

}
