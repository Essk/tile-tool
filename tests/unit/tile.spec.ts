import { Tile } from '@/utils/tile.ts';

describe('Tile', () => {
  it('instantiates', () => {
    const tile = new Tile({
      name: '',
    });
    expect(tile).toBeInstanceOf(Tile);
  });
  it('takes a name from passed options', () => {
    const tile = new Tile({name : 'a unique tile'});
    expect(tile.name).toBe('a unique tile');
  });
  it('gets a default id', () => {
    const tile = new Tile({name: ''});
    expect(tile.id).toBeTruthy();
  });
  it('can accept a paletteSet id', () => {
    const tile = new Tile ({paletteSet: 'dat_one'});
    expect(tile.paletteSet).toBe('dat_one');
  });
  it('defaults to pallete 0 of the palleteSet', () => {
    const tile = new Tile ({paletteSet: 'dat_one'});
    expect(tile.currentPalette).toBe(0);
  });
  it('accepts valid currentPalette indexes', () => {
    const tiles = [
      new Tile({currentPalette: 0, id: 'a'}),
      new Tile({currentPalette: 3, id: 'a'}),
      new Tile({currentPalette: 12, id: 'a'}),
      new Tile({currentPalette: 90, id: 'a'}),
      new Tile({currentPalette: -2, id: 'a'}),
      new Tile({currentPalette: 'three', id: 'a'}),
      new Tile({currentPalette: 2.6, id: 'a'}),
    ];
    expect(tiles).toMatchSnapshot();
  });
  it('can optionally accept an id from parameters', () => {
    const tile = new Tile({id: 'badgersnuff'});
    expect(tile.id).toBe('badgersnuff');
  });
  it('can accepy a partial pixel array', () => {
    const px = [15, 0, 7, 3];
    const tile = new Tile ({ px });
    expect(tile.px).toMatchSnapshot();
  });
  it('has a pixel array of the correct length of 0s', () => {
    const tile = new Tile({name: ''});
    expect(tile.px).toMatchSnapshot();
  });
  it('updates valid pixel data', () => {
    const tile = new Tile({name: ''});
    tile.updatePx(3, 12);
    expect(tile.px).toMatchSnapshot();
  });
  it('rejects invalid pixel data', () => {
    const tile = new Tile({name: ''});
    tile.updatePx(-1, 12);
    tile.updatePx(65, 12);
    tile.updatePx(5, -3);
    tile.updatePx(5, 2.4);
    tile.updatePx(5, -3);
    tile.updatePx(3.6, -3);
    expect(tile.px).toMatchSnapshot();
  });
});
