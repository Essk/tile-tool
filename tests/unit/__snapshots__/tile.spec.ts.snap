// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile accepts valid currentPalette indexes 1`] = `
Array [
  Tile {
    "_currentPalette": 0,
    "_id": "a",
    "_len": 64,
    "_name": "new tile",
    "_paletteSet": "",
    "_px": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  },
  Tile {
    "_currentPalette": 3,
    "_id": "a",
    "_len": 64,
    "_name": "new tile",
    "_paletteSet": "",
    "_px": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  },
  Tile {
    "_currentPalette": 12,
    "_id": "a",
    "_len": 64,
    "_name": "new tile",
    "_paletteSet": "",
    "_px": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  },
  Tile {
    "_currentPalette": 0,
    "_id": "a",
    "_len": 64,
    "_name": "new tile",
    "_paletteSet": "",
    "_px": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  },
  Tile {
    "_currentPalette": 0,
    "_id": "a",
    "_len": 64,
    "_name": "new tile",
    "_paletteSet": "",
    "_px": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  },
  Tile {
    "_currentPalette": 0,
    "_id": "a",
    "_len": 64,
    "_name": "new tile",
    "_paletteSet": "",
    "_px": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  },
  Tile {
    "_currentPalette": 0,
    "_id": "a",
    "_len": 64,
    "_name": "new tile",
    "_paletteSet": "",
    "_px": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  },
]
`;

exports[`Tile can accepy a partial pixel array 1`] = `
Array [
  15,
  0,
  7,
  3,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`Tile has a pixel array of the correct length of 0s 1`] = `
Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`Tile rejects invalid pixel data 1`] = `
Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`Tile updates valid pixel data 1`] = `
Array [
  0,
  0,
  0,
  12,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;
